### PR TITLE
Rewrite explanation, refactor CPP and (gonna) add Java for Distinct Values Queries (CSES1734)

### PR DESCRIPTION
**Pull request by ChuangSheep**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

File: solutions/cses-1734.mdx [hunk 1]
```diff
@@ -15,61 +15,196 @@ Sort the queries in a specific order before answering them.
 
 **Time Complexity:** $\mathcal{O}(N\log{N})$
 
-Answer queries in order of the right indices. Use a BIT to keep track of indices
-that contain distinct values. For every value, update the latest index for it
-and answer all queries at the current index. This ensures later indices are
-included in the BIT to accurately answer queries.
+We answer the queries from right to left, sorted by their left indices. By using a BIT, we can keep track of indices that contain distinct values. For every value $arr[i]$, we set the latest (i.e. the left-most) index in $bit[i]$ to $1$. If this value appeared before, we set the last used index back to $0$. Then, we can answer all queries with $a == i$ by summing up the values between $[i, b]$. This ensures later indices after $i$ are included in the BIT to accurately answer queries.
```
Comment (SansPapyrus683) [inline solutions/cses-1734.mdx:18]: > $a==i$

or just $a=i$?

File: solutions/cses-1734.mdx [hunk 2]
```diff
@@ -15,61 +15,196 @@ Sort the queries in a specific order before answering them.
...
-map<int, int> fst;
-
-int qry (int i) {
-	int res = 0;
-	for (; i; i -= i&(-i)) {
-		res += bit[i];
+struct BIT {
+	int size;
+	vector<int> bit;
+	BIT(int n) : size(n), bit(n, 0) {}
```
Comment (SansPapyrus683) [inline solutions/cses-1734.mdx:30]:
```suggestion
	BIT(int n) : size(n), bit(n) {}
```

File: solutions/cses-1734.mdx [hunk 3]
```diff
@@ -15,61 +15,196 @@ Sort the queries in a specific order before answering them.
...
-int n, q, sol[MX];
-vector<pair<int, int>> query[MX];
-vector<int> x(MX);
-map<int, int> fst;
-
-int qry (int i) {
-	int res = 0;
-	for (; i; i -= i&(-i)) {
-		res += bit[i];
+struct BIT {
```
Comment (SansPapyrus683) [inline solutions/cses-1734.mdx:27]: wrap in code segment maybe

File: solutions/cses-1734.mdx [hunk 4]
```diff
@@ -15,61 +15,196 @@ Sort the queries in a specific order before answering them.
...
-	for (int i = n; i >= 1; i--) {
-		int z = x[i - 1];
-		if (fst.count(z)) upd(fst[z], -1);
-		fst[z] = i;
-		upd(i, 1);
-		for (auto t : query[i]) {
-			sol[t.second] = qry(t.first);
+
+	// BIT also has to be 1-indexed, or the bitwise operator won't work.
+	BIT bit(n + 1);
```
Comment (SansPapyrus683) [inline solutions/cses-1734.mdx:63]: if the driver code has to do some weird crap to make sure the class works, then the class is implemented wrong
i'd put the +1 in the constructor (see the bit code i wrote for ref)

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

**Review by SansPapyrus683 — APPROVED** [review]
ily :heart:

**Review by ryanchou-dev — APPROVED** [review]
yes!

**Pull request merged**